### PR TITLE
Added DependencyInstaller to copy dependencies from s3 to maven local

### LINKED ISSUE #276: Setup maven and build dependencies using DependencyInstaller
Required for 
* Integ tests #327 
* BwC tests

### PR DESCRIPTION
**Pull request by VachaShah**
Signed-off-by: Vacha <vachshah@amazon.com>

### Description
This PR introduces a DependencyInstaller, requirements in [#276](https://github.com/opensearch-project/opensearch-build/issues/276#issuecomment-910800754). This will be used by integration tests, performance tests and bwc tests when run in the pipeline. It covers the following scenarios:
1. Install maven dependencies from S3 to maven local
2. Download build dependencies from S3 to a custom local path (to be used for scenarios like [this](https://github.com/opensearch-project/opensearch-build/blob/main/tools/standard-test/integtest_dependencies_opensearch.sh))
3. Clean up the downloaded contents

Note: I have merged the changes from #367 to use the S3 utility functions.
 
### Issues Resolved
#276 
 
### Check List
- [X] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by dblock — DISMISSED** [review]

File: bundle-workflow/src/test_workflow/dependency_installer.py
Comment (setiah) [inline bundle-workflow/src/test_workflow/dependency_installer.py:10]: instead of "build_id, version, arch" -> use build_manfiest as input, which has all these and more. Also, you can skip dependency_name in __init__. DependencyInstaller should provide a method to sync  - all maven dependencies, or - a list of dependencies.

File: bundle-workflow/tests/test_workflow/helper/test_utils.py
Comment (peternied) [inline bundle-workflow/tests/test_workflow/helper/test_utils.py:5]: Could you declare this function inside of the test class, if you foresee its reuse elsewhere please add comments or additional configuration

File: bundle-workflow/src/test_workflow/dependency_installer.py
Comment (VachaShah) [reply bundle-workflow/src/test_workflow/dependency_installer.py:32]: This is for the TODO where the files will be downloaded from S3 once the functionality is ready in PR #316.

File: bundle-workflow/src/test_workflow/dependency_installer.py
Comment (VachaShah) [reply bundle-workflow/src/test_workflow/dependency_installer.py:10]: Do you mean that if the user passes in the list of dependencies, then get the files for those dependencies and if nothing is passed in, get the files for all dependencies?

File: bundle-workflow/src/test_workflow/dependency_installer.py
Comment (VachaShah) [reply bundle-workflow/src/test_workflow/dependency_installer.py:27]: The `pipenv run black .` reformatted it this way since the `flake8` was failing with the if section being right indented. Definitely agree on the high bar! For the final code, the currently stubbed method for downloading the dependencies from s3 needs to be replaced once that functionality is merged and the tests corresponding to that stubbed method would change as well. I raised this PR early to get it out there and get feedback on the other parts and then can get a final review once the stubbed method is replaced.

File: bundle-workflow/src/test_workflow/dependency_installer.py
Comment (setiah) [reply bundle-workflow/src/test_workflow/dependency_installer.py:16]: i'm not sure what you mean by workers reuse. What I'm saying here is the workers would sync the dependencies in ~/.m2/repository of jenkins host. If there are two concurrent test jobs running, one for 1.1 another for 1.0, since they filesystem is same, they might end up overwriting each others dependencies.

File: bundle-workflow/src/test_workflow/dependency_installer.py
Comment (setiah) [reply bundle-workflow/src/test_workflow/dependency_installer.py:10]: Yeah 2 functions essentially - 
1. for syncing a list of maven dependencies say for a specific plugin.
2. for syncing all maven dependencies from build/ folder.

Comment (setiah) [reply bundle-workflow/src/test_workflow/dependency_installer.py:10]: 2. can be treated optional for now. 1 is must have

File: bundle-workflow/src/test_workflow/dependency_installer.py
Comment (VachaShah) [reply bundle-workflow/src/test_workflow/dependency_installer.py]: Added build-id in the path to keep the executions independent.

File: bundle-workflow/src/test_workflow/dependency_installer.py
Comment (VachaShah) [reply bundle-workflow/src/test_workflow/dependency_installer.py]: Added functionality for 1.

File: bundle-workflow/tests/test_workflow/helper/test_utils.py
Comment (VachaShah) [reply bundle-workflow/tests/test_workflow/helper/test_utils.py]: Removed the common test utils and added it in the test class itself.

**Review by dblock — CHANGES_REQUESTED** [review]
This is good! I think there are some should/must have's below but it's close.

File: bundle-workflow/tests/tests_test_workflow/test_dependency_installer.py
Comment (dblock) [inline bundle-workflow/tests/tests_test_workflow/test_dependency_installer.py:62]: This only makes sure "something" is returned. Check that len(), the number of files returned is > 0.

**Review by dblock — APPROVED** [review]

**Review by dblock — APPROVED** [review]

File: bundle-workflow/src/test_workflow/dependency_installer.py
Comment (setiah) [inline bundle-workflow/src/test_workflow/dependency_installer.py:17]: we can use the `download_folder` method available in #367

File: bundle-workflow/src/test_workflow/dependency_installer.py
Comment (setiah) [reply bundle-workflow/src/test_workflow/dependency_installer.py:5]: Dependency installer requirements - https://github.com/opensearch-project/opensearch-build/issues/276#issuecomment-910800754

**Review by peternied — APPROVED** [review]

**Comment by VachaShah** [comment]
Hi @dblock, sorry for a review request again for this PR, just updated the installer with the requirements mentioned in [#276 (comment)](https://github.com/opensearch-project/opensearch-build/issues/276#issuecomment-910800754) to handle more use cases and merged it with Himanshu's PR for S3 utility to use that functionality. Would love your review on it!

File: bundle-workflow/src/s3_utility.py
Comment (dblock) [inline bundle-workflow/src/s3_utility.py:1]: I had made a ton of comments on this file in a different PR, I think in #367 but I am no longer sure. Those needs to be addressed first.

**Review by dblock — CHANGES_REQUESTED** [review]
I had made a ton of comments on this file in a different PR, I think in #367 but I am no longer sure. Those needs to be addressed first.

File: bundle-workflow/src/s3_utility.py
Comment (VachaShah) [reply bundle-workflow/src/s3_utility.py:1]: Yes that is the PR that introduces the S3 utility. I merged it in my branch to call the download_folder method from #367, so it is showing up in my PR.

Comment (dblock) [reply bundle-workflow/src/s3_utility.py:1]: Ok, that code is not ready to be merged anywhere.

File: bundle-workflow/src/test_workflow/dependency_installer.py
Comment (setiah) [inline bundle-workflow/src/test_workflow/dependency_installer.py:31]: Could you help me understand how the dependency list is expected here? Seems like it expects the paths instead of dependency names. I think we should accept dependency names and generate paths to pull down from S3.

File: bundle-workflow/src/test_workflow/dependency_installer.py
Comment (setiah) [inline bundle-workflow/src/test_workflow/dependency_installer.py:44]: it should be list of dependency names

Comment (setiah) [reply bundle-workflow/src/test_workflow/dependency_installer.py:44]: need to change the code to handle this

File: bundle-workflow/src/test_workflow/dependency_installer.py
Comment (setiah) [inline bundle-workflow/src/test_workflow/dependency_installer.py:51]: The paths where these dependencies are installed, should be members of the DependencyInstaller object. The caller need not provide them, it can be optional though. DependencyInstaller should be able to cleanup everything it created by itself w/o input requirements from caller

File: bundle-workflow/src/test_workflow/dependency_installer.py
Comment (setiah) [inline bundle-workflow/src/test_workflow/dependency_installer.py:39]: save the custom_local_path to cleanup later in the cleanUp method

File: bundle-workflow/src/test_workflow/dependency_installer.py
Comment (VachaShah) [reply bundle-workflow/src/test_workflow/dependency_installer.py:31]: It is the dependency name that is expected here. [Example tests](https://github.com/opensearch-project/opensearch-build/pull/318/files#diff-05f802ba74b1ca82bd5f6ef6241b4a5bc5a71035c756ff71b53ae963690475b1R22)

File: bundle-workflow/src/test_workflow/dependency_installer.py
Comment (VachaShah) [reply bundle-workflow/src/test_workflow/dependency_installer.py:44]: I will edit the comment to dependency names.

File: bundle-workflow/src/s3_utility.py
Comment (setiah) [inline bundle-workflow/src/s3_utility.py:7]: You can untrack these files in your commit and still use them for local development. That way, they won't be raised as part of your PR.

File: bundle-workflow/src/test_workflow/dependency_installer.py
Comment (VachaShah) [reply bundle-workflow/src/test_workflow/dependency_installer.py:51]: The custom path for build dependencies would have to be passed by the user, for maven dependencies I can try to store the location to use for cleanup.

**Comment by VachaShah** [comment]
This PR is dependent on the S3 utility PR #367. I will raise the final version once that PR is merged in.

File: bundle-workflow/tests/tests_test_workflow/test_dependency_installer.py
Comment (dblock) [inline bundle-workflow/tests/tests_test_workflow/test_dependency_installer.py:62]: Don't use camelCase in Python.

File: bundle-workflow/src/test_workflow/dependency_installer.py
Comment (setiah) [inline bundle-workflow/src/test_workflow/dependency_installer.py:22]: nit: this location provider logic is common across build and test workflows and should be read from a single place. Check if there's already a utility/or we can make it. It's okay if you create an issue and take it separately outside this PR.

File: bundle-workflow/src/test_workflow/dependency_installer.py
Comment (setiah) [reply bundle-workflow/src/test_workflow/dependency_installer.py:31]: If that's the case, this path seems wrong ` s3_path = f"{self.s3_maven_location}/{dependency}"` It exists under org/opensearch...

File: bundle-workflow/src/test_workflow/dependency_installer.py
Comment (setiah) [reply bundle-workflow/src/test_workflow/dependency_installer.py:51]: thx, that works.

File: bundle-workflow/Pipfile
Comment (owaiskazi19) [inline bundle-workflow/Pipfile:17]: Instead of this, we can install it like `boto3-stubs[s3]` and then later import `from mypy_boto3_s3 import S3Client, S3ServiceResource`.  We won't need `boto3-stubs` then if we are using boto3 only for S3.
Ref: https://pypi.org/project/mypy-boto3-s3/

File: bundle-workflow/src/test_workflow/dependency_installer.py
Comment (VachaShah) [reply bundle-workflow/src/test_workflow/dependency_installer.py]: For maven dependencies, each dependency would have a separate path, if we store them when the install_maven_dependencies is called, we would end up cleaning all of them in the cleanup method. Would it not be better if the user passes in the path that needs to be cleaned so that there is no incorrect cleanup? LMK what you think.

For all the other comments and requirements, I have made the changes accordingly.

File: bundle-workflow/src/test_workflow/dependency_installer.py
Comment (dblock) [inline bundle-workflow/src/test_workflow/dependency_installer.py:64]: Use `logging.error`.

File: bundle-workflow/src/test_workflow/dependency_installer.py
Comment (dblock) [inline bundle-workflow/src/test_workflow/dependency_installer.py:54]: This deletes a folder, which is ... dangerous if we get it wrong. For maven dependencies I don't think we will ever want to be deleting them after downloading explicitly because they will be intermixed with other dependencies that may have been potentially dropped into `~m2`. We already are doing cleanup in https://github.com/opensearch-project/opensearch-build/pull/403. I think you should delete it from the PR and close any issues that say otherwise.

File: bundle-workflow/tests/tests_test_workflow/test_dependency_installer.py
Comment (dblock) [inline bundle-workflow/tests/tests_test_workflow/test_dependency_installer.py:54]: = S3Bucket() - because it's mocked you will get the mock object

File: bundle-workflow/tests/tests_test_workflow/test_dependency_installer.py
Comment (dblock) [inline bundle-workflow/tests/tests_test_workflow/test_dependency_installer.py:21]: = S3Bucket()

File: bundle-workflow/tests/tests_test_workflow/test_dependency_installer.py
Comment (dblock) [inline bundle-workflow/tests/tests_test_workflow/test_dependency_installer.py:22]: This is a test, don't assign anything to `self.` other than in `setUp`. Looks like `dependency_installer` is needed in every test, so move this into `setUp`.

File: bundle-workflow/tests/tests_test_workflow/test_dependency_installer.py
Comment (dblock) [inline bundle-workflow/tests/tests_test_workflow/test_dependency_installer.py:31]: Extract into a method in `DependencyInstaller` that gives you a maven local path for a dependency. You have similar code in DependencyInstaller itself.

File: bundle-workflow/tests/tests_test_workflow/test_dependency_installer.py
Comment (dblock) [inline bundle-workflow/tests/tests_test_workflow/test_dependency_installer.py:43]: This will become something like `dependency_installer.maven_local_path('opensearch-job-scheduler')`

**Review by dblock — APPROVED** [review]
I have some should have's, but won't hold the PR for them.

File: bundle-workflow/src/test_workflow/dependency_installer.py
Comment (setiah) [inline bundle-workflow/src/test_workflow/dependency_installer.py:38]: nit: this is s3_relative_path

File: bundle-workflow/src/test_workflow/dependency_installer.py
Comment (setiah) [inline bundle-workflow/src/test_workflow/dependency_installer.py:31]: This looks incorrect. S3Bucket needs to be initialized with bucket_name

File: bundle-workflow/src/test_workflow/dependency_installer.py
Comment (setiah) [reply bundle-workflow/src/test_workflow/dependency_installer.py:38]: It should be s3_maven_location for the dependency  which is a relative path inside the bucket. An example -  `builds/1.1.0/24/x64/maven/org/opensearch/`

File: bundle-workflow/src/test_workflow/dependency_installer.py
Comment (setiah) [reply bundle-workflow/src/test_workflow/dependency_installer.py:54]: @dblock - valid point. But I think we still need to cleanup dependencies which are synced as part of one test run, else we will unnecessarily bloat the local maven over a period of time. IMO we should switch back to the custom maven local path, based on a random tmp location, where we download dependencies and cleanup after the test run. This keeps the concurrent test runs isolated. We planned on doing it earlier but then switched to having docker provide that isolation by mounting on a randomized tmp path. I think we need it in the code as developers will run this code on local setup as well where docker mounting won't provide that sort of isolation. @VachaShah and I discussed about it yesterday.

File: bundle-workflow/tests/tests_test_workflow/test_dependency_installer.py
Comment (setiah) [inline bundle-workflow/tests/tests_test_workflow/test_dependency_installer.py:19]: you can replace this with S3Bucket that's checked into main

File: bundle-workflow/src/test_workflow/dependency_installer.py
Comment (peternied) [reply bundle-workflow/src/test_workflow/dependency_installer.py:54]: >bloat the local maven over a period of time.

Jenkins Agent nodes will only exist for a single workflow then are torn down, how fast is this bloat?  Enough that we need to clean it out to complete a workflow?

Comment (dblock) [reply bundle-workflow/src/test_workflow/dependency_installer.py:54]: For the record CI will be providing throw away containers with their own file system for these jobs.

File: bundle-workflow/tests/tests_test_workflow/test_dependency_installer.py
Comment (VachaShah) [reply bundle-workflow/tests/tests_test_workflow/test_dependency_installer.py:46]: I tried it but it tries to call the actual constructor instead of mocked.

File: bundle-workflow/tests/tests_test_workflow/test_dependency_installer.py
Comment (VachaShah) [reply bundle-workflow/tests/tests_test_workflow/test_dependency_installer.py:20]: This is the mocked version of the `S3Bucket` checked in main.

File: bundle-workflow/src/test_workflow/dependency_installer.py
Comment (VachaShah) [reply bundle-workflow/src/test_workflow/dependency_installer.py]: Removed the cleanup logic.

**Comment by VachaShah** [comment]
Hi @dblock and @setiah, I have addressed the comments. LMK if this looks good to be merged.

**Review by dblock — APPROVED** [review]

**Review by peternied — APPROVED** [review]

**Review by setiah — APPROVED** [review]
Thanks @VachaShah. LGTM!

**Pull request merged**
